File: hotel.rb
require("date")

class Hotel
  attr_reader :id, :name, :rooms

  @@id = 0
  def initialize(name)
    @id = @@id += 1
    @name = name
  end

  def rooms
    Room.select { _1.hotel == self }
  end

  def rooms_available_between(start_date, end_date)
    rooms.reject { _1.booked_between?(start_date, end_date) }
  end
end

class Room
  attr_reader :id, :hotel

  @@id = 0
  def initialize(hotel)
    @id = @@id += 1
    @hotel = hotel
  end

  def book(start_date, end_date)
    return if booked_between?(start_date, end_date)

    Booking.new(room: self, start_date: start_date, end_date: end_date)
  end

  def booked_on?(date)
    bookings.any? { _1.period.include?(date) }
  end

  def booked_between?(start_date, end_date)
    (start_date...end_date).any? { booked_on?(_1) }
  end

  def bookings
    Booking.all.select { _1.room  == self }
  end
end

class Booking
  attr_reader :id, :room, :start_date, :end_date

  @@id = 0
  @@bookings = []

  def self.all
    @@bookings
  end

  def initialize(room:, start_date:, end_date:)
    @id = @@id += 1
    @room = room
    @start_date = start_date
    @end_date = end_date
    @@bookings << self
  end

  def period
    (start_date..end_date)
  end

end
#  require("./hotel")
hotel = Hotel.new("Ritz")
room1 = Room.new(hotel)
room2 = Room.new(hotel)

room1.book(Date.new(2021, 1, 1), Date.new(2021, 1, 10))
room1.book(Date.new(2021, 1, 20), Date.new(2021, 1, 30))

room1.booked_on?(Date.new(2021, 1, 5)) # true
room1.booked_on?(Date.new(2021, 1, 15)) # false

room1.booked_between?(Date.new(2021, 1, 5), Date.new(2021, 1, 7)) # true
room1.booked_between?(Date.new(2021, 1, 5), Date.new(2021, 1, 15)) # true
room1.booked_between?(Date.new(2021, 1, 12), Date.new(2021, 1, 15)) # false
room1.booked_between?(Date.new(2021, 1, 15), Date.new(2021, 2, 1)) # true

File: hotel.py
from datetime import datetime, timedelta
from datetimerange import DateTimeRange

class Hotel:
    id = 0
    def __init__(self, name):
        self.name = name
        Hotel.id += 1
        self.id = Hotel.id

    def __str__(self):
        return f"name: {self.name}"
    
    def __repr__(self):
        return f"hotel<name: {self.name}>"

    def rooms(self):
        return [room for room in Room.rooms if room.hotel == self]

    def rooms_available(self, start, end):
        return [room for room in self.rooms() if not room.is_booked_between(start, end)]


# class RoomRepository(object):
#     _instance = None

#     def __new__(cls):
#         if cls._instance is None:
#             cls._instance = super(RoomRepository, cls).__new__(cls)
#         return cls._instance
    
#     def __init__(self):
#         self.current_id = 0
#         self.rooms = []
    
#     def add_room(self, room):
#         self.current_id += 1
#         room.id = self.current_id
#         self.current_id += 1
#         self.rooms.append(room)  

class Room:
    id = 0
    rooms = []

    @classmethod
    def all(cls):
      return cls.rooms

    def __init__(self, hotel):
        self.hotel = hotel
        Room.id += 1
        self.id = Room.id
        Room.rooms.append(self)
    
    def bookings(self):
        return [booking for booking in Booking.bookings if booking.room == self]
  
    def book(self, start, end):
        if self.is_booked_between(start, end):
            raise ValueError("Room is already booked for this date range")
        return Booking(self, start, end)
  
    def is_booked_between(self, start, end):
        date_range = DateTimeRange(start, end).range(timedelta(days=1))
        return any(
          self.is_booked_on(date) for date in date_range
        )
    
    def is_booked_on(self, date):
        return any(
          date in booking.period for booking in self.bookings()
        )

    def __str__(self):
        return f"hotel: {self.hotel} id: {self.id}"

    def __repr__(self):
        return f"room<hotel: {self.hotel} id: {self.id}>"


class Booking:
    id = 0
    bookings = []

    def __init__(self, room, start, end):
        self.room = room
        self.start = start
        self.end = end
        self.period = DateTimeRange(self.start, self.end)
        Booking.id += 1
        self.id = Booking.id
        Booking.bookings.append(self)
